### PR TITLE
Support narrowcast message API

### PR DESCRIPTION
**Pull request by clsung**
- implments #190
- add field `RequestID` to BasicResponse
TODO:
- narrowcast progress API

### CODE REVIEW

File: linebot/filter.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package linebot
+
+import "encoding/json"
+
+// Filter interface
+type Filter interface {
+	Filter()
```
Comment (nasa9084) [inline linebot/filter.go]: It seems Filter() method do nothing, just represents the struct is one of filter.
If so, I think it is better to write to comment, and IMHO, the better method signature is:
```
IsFilter() bool
```
and each implementation of this method should return true.
With just `Filter()`, it seems we can filtering with the method.
HDYT?

File: linebot/filter.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+// AudienceFilter type
+type AudienceFilter struct {
```
Comment (nasa9084) [inline linebot/filter.go]: The Audience object is not an demographic filter object, it is a recipient object.
Currently recipient object and demographic filter object have (looks) same operator object but they are not exactly same, they can be altered for each.
So I think it is better to split operator struct to two, 1. recipient operator object 2. demographic filter operator object.

File: linebot/filter.go (hunk 1 above)
Comment (clsung) [reply linebot/filter.go]: Hi, I think it's ok to change to IsFilter(), however the intention is to comply with the naming rule, i.e. 
`// Message interface
type Message interface {
    Message()
}`

I have no objection on this. Just need a decision :)

File: linebot/filter.go (hunk 2 above)
Comment (clsung) [reply linebot/filter.go]: That's a good idea, will update later.

File: linebot/filter.go (hunk 1 above)
Comment (nasa9084) [reply linebot/filter.go]: ah.. sorry, I haven't see that interface, your naming is right.
Thanks,

File: linebot/filter.go (hunk 2 above)
Comment (clsung) [reply linebot/filter.go]: Hi, I've renamed `filter` to `selector` to remedy the semantics.  Thanks.

**Comment by nasa9084** [comment]
I think interfaces for audience and other filters should be separately defined, such as:

```
type Recipient interface {
   // method
}

type DemographicFilter interface {
  // method
}
```

and Audience implements Recipient and other demographic filter objects implements DemographicFilter

sorry my lack of explanation

**Comment by oklahomer** [comment]
Since recipient and demographic filter are entirely different objects, can we treat them as such?
The current implementation shares the same interface such as Selector, but I think it should be separated as they have no common ancestry.
And hence, the `Operator` itself will be separated for both as a result since this is not meant to be shared.

**Comment by clsung** [comment]
> Since recipient and demographic filter are entirely different objects, can we treat them as such?
> The current implementation shares the same interface such as Selector, but I think it should be separated as they have no common ancestry.
> And hence, the `Operator` itself will be separated for both as a result since this is not meant to be shared.

Hi, interface in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here. On the other hand, a type can implement multiple interfaces. After look again into Operator/Selector code, I think I should name it as Select() method and Selector interface in order to comply with Go coding guideline. 

How do you think?

**Comment by oklahomer** [comment]
My apologies for the interruption and confusion.

I thought the point was to separate Operator to simplify the interface/implementation: One operator for Recipient and another for Demographic filter.

The  [Recipient](https://developers.line.biz/en/reference/messaging-api/#narrowcast-recipient)  document says the operator deals with `Recipient` objects while [Demographic filter](https://developers.line.biz/en/reference/messaging-api/#narrowcast-demographic-filter)'s operator deals with `Demographic filter` objects. If we simply follow those definitions, we would have two different operators: `RecipientOperator` and `DemographicFilterOperator`. In this way, the interface and its implementation of the current `Selector` interface would also be distinguished accordingly and the overall design could be more simplified and type-safe, I suppose.

**Comment by clsung** [comment]
> My apologies for the interruption and confusion.
> 
> I thought the point was to separate Operator to simplify the interface/implementation: One operator for Recipient and another for Demographic filter.
> 
> The [Recipient](https://developers.line.biz/en/reference/messaging-api/#narrowcast-recipient) document says the operator deals with `Recipient` objects while [Demographic filter](https://developers.line.biz/en/reference/messaging-api/#narrowcast-demographic-filter)'s operator deals with `Demographic filter` objects. If we simply follow those definitions, we would have two different operators: `RecipientOperator` and `DemographicFilterOperator`. In this way, the interface and its implementation of the current `Selector` interface would also be distinguished accordingly and the overall design could be more simplified and type-safe, I suppose.

Ok, I think it's fine to separate them. Updated.

**Comment by nasa9084** [comment]
LGTM,

**Pull request merged**